spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py

```diff
@@ -286,28 +450,556 @@ def get_local_provenance_data(self):
+                    local_host=local_ip, remote_host=remote_ip)
+
+                # if we have not received confirmation of finish, try to
+                # retransmit missing seq nums
+                if not received_confirmation:
+                    self._outgoing_retransmit_missing_seq_nums(data_to_write)
+
+    def _read_in_missing_seq_nums(self, data, data_to_write, position):
+        """ handles a missing seq num packet from spinnaker
+
+        :param data: the data to translate into missing seq nums
+        :param data_to_write: the data to write
+        :param position: the position in the data to write.
+        :rtype: None
+        """
+        # find how many elements are in this packet
+        n_elements = (len(data) - position) / self.LENGTH_OF_DATA_SIZE
+
+        # store missing
+        self._missing_seq_nums_data_in[-1].extend(struct.unpack_from(
+            "<{}I".format(n_elements), data, position))
+
+        # determine if last element is end flag
+        if self._missing_seq_nums_data_in[-1][-1] == \
+                self.MISSING_SEQ_NUMS_END_FLAG:
+            del self._missing_seq_nums_data_in[-1][-1]
+            self._outgoing_retransmit_missing_seq_nums(data_to_write)
+        if (self._total_expected_missing_seq_packets == 0 and
+                self._have_received_missing_seq_count_packet):
+            self._outgoing_retransmit_missing_seq_nums(data_to_write)
+
+    def _outgoing_process_packet(self, data, data_to_write):
+        """ processes a packet from SpiNNaker
+
+        :param data: the packet data
+        :param data_to_write: the data to write to spinnaker
+        :return: if the packet contains a confirmation of complete
+        :rtype: bool
+        """
+        position = 0
+        command_id = struct.unpack_from("<I", data, 0)[0]
+        position += self.COMMAND_ID_SIZE
+        # log.info("received packet with id {}\n".format(command_id))
+
+        # process first missing
+        if command_id == \
+                self.SDP_PACKET_RECEIVE_FIRST_MISSING_SEQ_DATA_IN_COMMAND_ID:
+
+            # find total missing
+            self._total_expected_missing_seq_packets += \
+                struct.unpack_from("<I", data, position)[0]
+            position += self.MISSING_SEQ_PACKET_COUNT_SIZE
+            self._have_received_missing_seq_count_packet = True
+
+            # write missing seq nums and retransmit if needed
+            self._read_in_missing_seq_nums(data, data_to_write, position)
+
+        # process missing seq packets
+        if command_id == \
+                self.SDP_PACKET_RECEIVE_MISSING_SEQ_DATA_IN_COMMAND_ID:
+            # write missing seq nums and retransmit if needed
+            self._total_expected_missing_seq_packets -= 1
+
+            self._read_in_missing_seq_nums(data, data_to_write, position)
+
+        # process the confirmation of all data received
+        if command_id == self.SDP_PACKET_RECEIVE_FINISHED_DATA_IN_COMMAND_ID:
```

Comment:
Another `if boolean: return True else: return False`; John Latham would be proud!

Reply:
fixed. alas no metaphorical gun on the table